### ISSUE TITLE
Customizing color mapping while using domain that comes from data only

### ISSUE THREAD

**Issue by jowens** [vega/vega#2535]
Is it possible to specify a domain/range for a scale and then have only the items that are actually in the data appear in the legend? Right now if I put 6 labels in the domain/range and only 3 appear in the data, all 6 still appear in the legend.

The reason I do this is because I have a bunch of plots and I want the colors consistent between the plots, even though some plots have all the labels and some only have a subset.

More context here: https://github.com/altair-viz/altair/issues/2075

```json
{
  "config": {"view": {"continuousWidth": 400, "continuousHeight": 300}},
  "data": {"name": "data-9794decafd64313e2c537fe6fc95bda8"},
  "mark": "point",
  "encoding": {
    "color": {
      "type": "nominal",
      "field": "label",
      "scale": {
        "domain": ["A", "B", "C"],
        "range": ["#4daf4a", "#984ea3", "#ff7f00"]
      }
    },
    "shape": {"type": "nominal", "field": "label"},
    "x": {"type": "quantitative", "field": "x"},
    "y": {"type": "quantitative", "field": "y"}
  },
  "$schema": "https://vega.github.io/schema/vega-lite/v4.8.1.json",
  "datasets": {
    "data-9794decafd64313e2c537fe6fc95bda8": [
      {"x": 1, "y": 3, "label": "A"},
      {"x": 2, "y": 2, "label": "B"},
      {"x": 3, "y": 1, "label": "A"}
    ]
  }
}
```

**Comment by jheer** [comment]
You can do this in Vega with a little bit of data organization. For example, by:

1. Storing your desired domain/range values as their own data table.
2. Using a [lookup transform](https://vega.github.io/vega/docs/transforms/lookup/) to extract range values only for domain values that occur in the data.
3. Use a [range based on a data reference](https://vega.github.io/vega/docs/scales/#range), which ordinal scales support. Make sure to sort by the domain value so that the mapping doesn't get shuffled.

Here is [an example spec](https://vega.github.io/editor/#/url/vega/N4KABGBEAkDODGALApgWwIaQFxUQFzwAdYsB6UgN2QHN0A6agSz0QFcAjOxge1IRQyUa6SgFY6AK1jcAdpAA04KO3TwA1tQBO3VjIAm2KAHdEzZAqWRC6PXsYzqh0YoiQjjPS0MAWAAy+XXGRGanxDAGZ-QMhYPABPABtzHEh4ZASEi1c9dDxMHABtJQhQCDKoGXRUZKgcvIBaAE4AdkbvPWR4dAAzPQA2b3CARnDkACZ4UXDm7uQ+7vhG0XYcgA4s8qgKdATWZFhDIs2y4EgAD0Mx+Sg4y+vIBPR2dMNIACFIAF9A49OLnCG91uOHC90ez0yKQAgl8fps-hEgYZAVBwS9oV9imUALpY75Y0qbSCVaqvOroAD6vg25RiOk0aTJuXQTVa7U6PX6gxG40m01m80WyzWNLKkDwmnQMlg3W4mlQhyx5UJx1c8UINUg3UYCTwyE0ouOkGQZ0IBpSjFgADUdh4ABR1VioAoAHXObuxAEowAAyH1gS0AMXsZjtAGpHc63WcPd6-QHrba9A7ck7XZA4rHff6gyG9eHI+nM5AvZAlWVPuXceV8cry8SqpqMIRCPZHHDXNtdvtFaqwKc9NwMPZXjD7pKHJqAMTtHreTC11UDofoEcpD7jqXUaeNVbeZDocKw8slSCD4dyFIAYQUUAn29eU+63Rm-kxx2rFbhKrFJKb6BbNsKW1XV9WQAwOygaRWAZTVySpQ0oAlKUZTlBVChPftMLVOINVedBqC0YQ9UQ2ktB0Qh2GBMAClObV0gMFI0Uye50AOFJz1XORPk-Y5F1+bCkNwzUEm4bg1FYQhSLFbptHQqBm1bBxb0gNRkGos8VxHSDaXohI9HYmjNIvEsdLFLs9kMgpIHvcxeM2SsPzxJRsWiDBNDUKyCXrP9XnczzEPFYTXlgOJUHYbgWPrWJEhqazCG4ew8FM+tZKHQwB2ZJk8gQ-ixWQGR4G4DoMswyBJLqGofyNbhrHgZhqNOCyal8Ohmjyo0QMhftIGa15kOlaxNAK5KOqJWJtDUjKYi6JJXiK0SDXuPTGNRJ4XjG2l-h6hAdk1C5lsYBjXguTaxUama9teW5DuOlJbjO1xYEQACqsuuaUme16VJW15mPfVVHIc5yIFcyxdqSLy6yNXyUgO+t1RE+wDyW+tOLXHryWyylqVu-STuPI1bMOAIdpCSpurcDwvB4yDiUYRkcAlPY6YAL31bhDGZ5A8W-HzG2uwLEb+5H0FRo10cvTGso45kELx1aM0JolicKU5YHJnZXhQEIwm+MBfDBmGGZqbm2Y5rnND2XnvJhgWUgWuUheClI5TsCnAsl6asdlnLcagX6mPWlioLlZKmat5AxpsrcqrKn2FIApTqGAnU9WGiCA6O-GUmJsy6U0cOetqvztKzu7ai07jyyBiA8uqip7agl68Lp4XXc0d2tbpr2cEyvJsfl8uc7WiEVOkQvLb2aPVagsKIsyEGwCNqB0DOHsMOh8bZv2wK5SOmQi8gCKCHSumtA8ABlHfBfPzQPCn5A24ZtQrx0Q-po16gPYd4IEjtdwnhECkD8J6ZWtJe5gG6DsWAT96zMUMtAhIsC6YYDOAAUTOHqD+OAAj1lQPYTB2Ci54KNHgF+iCYFwKNKzewHRtq+BtlvWkENNQ3XrPvEaf1kDdGSnfK+N84aBQvqtM2CMX5v10EXdWmtKZpB1HaHWoQ8AgN8GA6OkCkEoPgcHShyDqFEjQUQrhuDUGEKwSYg2z91B6O0TQuhJpDCMJrHzI0rCCZ004Tg5Q3BT4Kn4atLRBjaTkLwB9KA8MjTMUDLsZ6j86bMQAPJUE0I8KSEcWbiPUJI7xMjv5a1-gowBLBVHqLZg4hhTCTjRUETcPe99LEPB4Xw+sIjDBBOfmEthgUkkpLSfErJr937SJiLI+af9FHBGUaU8BYpaH6EcbgpeK8mnbn0FDapbiJTiU1I7cW4156RQACouygPVBkc1OnhIeMHQKBUiolT7mVUK4VIqGQbrSCquQ459l0jqbqTUdh7H6hOWAQ0uGPTFLVVQDVpp9VwW1WugNMJIrrlUko0UW6ai+q3aKhyEgnLwg7RgFzzBXJErcum9zio-NVDEfF7zBLlUIJVUqvyZL-LhUCzUA0wViwhWZWk0L6rxC5d2JxiLBKoorDXZZIBPhAA).

@kanitw, as this is expressible in Vega, maybe it make sense to move the issue back to VL?

**Comment by nsonnad** [comment]
Hello, I'm running into this same issue (four years later). I'd like colors to be consistent across multiple charts, even if certain categories/groups are absent in certain charts, without absent groups showing up in the legend. Is there any way to express this in Vega-Lite as of now?

**Comment by PBI-David** [comment]
Yes, use the technique here to drive the colour from your supplied data and then it will always be consistent. 

https://vega.github.io/vega-lite/docs/scale.html#example-setting-color-range-based-on-a-field

**Comment by nsonnad** [comment]
> Yes, use the technique here to drive the colour from your supplied data and then it will always be consistent.
> 
> https://vega.github.io/vega-lite/docs/scale.html#example-setting-color-range-based-on-a-field

Thank you! This approach worked well for my use case.
